0010-zkvm-keccak.md

```diff
@@ -0,0 +1,778 @@
+
+If the input was not previously expanded, the next step would be to expand all 25 words. Each word of 64 bits would have to be split into 4 parts of 16 real bits each. The expansion itself would be performed through the `Reset` lookup table. This step would require $4\times25=100$ lookups.
+
+There are two main modes: absorb and squeeze. Inside the absorb mode, there are two additional submodes called pad and root, which can as well happen at the same time. 
+
+
+- **Absorb mode**
+
+    In absorb mode, the first coefficient is set to 1: `absorb := coeff[0] = 1`, and the second coefficient is 0.
+
+    The gate takes a block as 68 chunks of 16 bits expanded, pads with zeros until reaching 100 chunks (1600 bits) to fit one whole state, and XORs it with the previous state.
+
+    The zero padding is enforced performing the following 32 constraints:
+
+    ```rust
+    constrain(absorb * zeros[i])
+    ```
+
+    The XOR part of the gate uses $100$ constraints 
+
+    ```rust
+    constrain(absorb * (xor_state[i] - (old_state[i] + new_state[i])))
+    ```
+    where `new_state` is the concatenation of `new_block` and `zeros`.
+
+    Additionally, the gate must check the decomposition into shifts of the new block.
+
+    ```rust
+    for i in [0..100) {
+        constrain(absorb * (new_block[i] - (
+                            shifts0(i)
+                            + 2 * shifts1(i)
+                            + 4 * shifts2(i)
+                            + 8 * shifts3(i) )))
+    }
+    ```
+
+    The correspondence between the shifts and the dense bytes will happen inside a lookup pattern.
+
+    - **Root mode**
+
+        In this mode, the third coefficient is set to 1: `root := coeff[2] = 1`, and `absorb = 1, squeeze = 0`. 
+
+        This mode is only performed once per hash, and it must happen in the first absorb. The only difference between this one and any other absorb is that the root mode sets the initial state to all zeros. For that reason, apart from the constraints above, it checks:
+
+        ```rust
+        constrain(root * old_state[i])
+        ```
+    
+    - **Pad mode** 
+
+        Padding with the $10^*1$ rule only happens once per hash. But unlike the root mode, padding happens only in the last absorb. That is because the input message is padded until reaching a length that is a multiple of 136 bytes, which always fully fits in one whole block. This means, in the case that the input message is $\leq 135$ bytes in length, the single absorb will run in both `root` and `pad` modes.
+
+        When in pad mode, we will have 136 selector flags (coefficient indices `[4..140)`) that will activate at most 136 positions corresponding to the padding. Additionally, each of the 136 coefficients corresponding to the bytes of the new block are set to either `0x00`, `0x01`, `0x80`, or `0x81`, which are all the possible combinations that a padding byte can take. In this mode, `absorb = 1, squeeze = 0`. 
```

Comment:
These being 'coefficients' means that they're statically knowned and fixed in the circuit definition, right? I don't think that's the case if we don't statically know the length of hashes ahead of time (and I believe we don't).
It seems like we need some way of constraining these dynamically, in response to the preimages that are requested.

Reply:
But this seems to be a problem not only of padding, but of block width as well. Meaning, if the input happens to take more than 135 bytes, then the circuit will need to have more rows as well. So there's no "universal" keccak circuit at this point. So, if you need to know the length in order to setup the circuit with X-many absorbs, why would the padding be a problem? The only difference relies on a more accurate input length (number of absorbs only needs to know what multiple of the block length is upperbounded by the length of the input; whereas the padding technique inside the sponge itself requires to know the exact byte length, not just an interval).